src/libcore/iter/range.rs

```diff
@@ -323,8 +577,16 @@ impl<A: Step> Iterator for ops::RangeFrom<A> {
 
     #[inline]
     fn nth(&mut self, n: usize) -> Option<A> {
-        let plus_n = self.start.add_usize(n).expect("overflow in RangeFrom::nth");
-        self.start = plus_n.add_one();
+        // If we would jump over the maximum value, panic immediately.
+        // This is consistent with behavior before the Step redesign,
+        // even though it's inconsistent with n `next` calls.
+        // To get consistent behavior, change it to use `forward` instead.
+        // This change should go through FCP separately to the redesign, so is for now left as a
+        // FIXME: make this consistent
```

Comment:
[scottmcm requested on a previous iteration of this patch that it doesn't change any observable behavior](https://github.com/rust-lang/rust/pull/62886#issuecomment-544311899). Also note that this isn't just about `nth`, but also `skip`.

The current behavior is [noted as unspecified by](https://doc.rust-lang.org/nightly/std/ops/struct.RangeFrom.html)

> *Note:* Currently, no overflow checking is done for the Iterator implementation; if you use an integer range and the integer overflows, it might panic in debug mode or create an endless loop in release mode. This overflow behavior might change in the future.

Personally, I think it's most correct to make sure that `nth`/`skip` are equivalent to the default implementation of just calling `next` multiple times. In any case, I agree with @scottmcm that it's better to leave the observable change to a separate PR so this PR can be scoped entirely just to redesigning the unstable trait rather than changing stable behavior as a side effect.

(This design is definitely biased towards fixing this towards debug-checked overflow, however. This `Step` doesn't expose any guaranteed saturating or wrapping options.)

Reply:
I still feel that we should go ahead and do this. I strongly doubt that this will be an issue in practice, and we can always revert it if it causes issues (though it would really mean that code is relying on unspecified behavior).

cc @rust-lang/libs

Reply:
@Amanieu to be clear, are you advocating for consistent `forward_checked` (panic on overflow) or consistent `forward` (debug checked overflow) here?

Reply:
I am advocating for `forward_checked`.

Reply:
Specifically I want the whole method to just forward directly to a single call of `forward_checked`.

Reply:
On second though, I can see why this is confusing, maybe we should leave it alone for now.

Reply:
I think that would straight up be unsound. `RangeFrom` is `TrustedLen` (disclaimer: `TrustedLen` is unstable), and the `size_hint` says `RangeFrom` is an unbounded iterator.